examples/computer_vision/byol_pytorch/model_def.py

```diff
@@ -0,0 +1,367 @@
+        self._init_transforms()
+        self._init_self_supervised()
+        self._init_classifiers()
+        self._init_reducers()
+        # Create a separate dataloader for training the classifier.
+        # With distributed training, calling get_data_loader on the Determined dataloader will automatically
+        # shard the dataset.
+        self.train_cls_dataloader = (
+            self._build_cls_training_data_loader().get_data_loader(
+                repeat=False,
+                num_replicas=self.context.distributed.get_size(),
+                rank=self.rank,
+            )
+        )
+
+    def _init_transforms(self) -> None:
+        """
+        Create training and evaluation transforms.
+        """
+        dataset_metadata = DATASET_METADATA_BY_NAME[self.data_config.dataset_name]
+        mean = dataset_metadata.mean
+        std = dataset_metadata.std
+        # BYOL paper uses two different distributions for training transforms.
+        self.data_config.train_transform_fn1 = build_training_transform(
+            self.data_config.train_transform1, mean, std
+        )
+        self.data_config.train_transform_fn2 = build_training_transform(
+            self.data_config.train_transform2, mean, std
+        )
+        self.data_config.eval_transform_fn = build_evaluation_transform(
+            self.data_config.eval_transform, mean, std
+        )
+
+    def _init_self_supervised(self) -> None:
+        """
+        Create BYOL network and optimizer.
+        """
+        backbone_metadata = BACKBONE_METADATA_BY_NAME[self.hparams.backbone_name]
+        net = backbone_metadata.build_fn()
+        # Transform random_crop_size determines final image size.
+        assert (
+            self.data_config.train_transform1.random_crop_size
+            == self.data_config.train_transform2.random_crop_size
+        ), "Crop size must be the same for all transforms."
+        assert (
+            self.data_config.eval_transform.center_crop_size
+            == self.data_config.train_transform1.random_crop_size
+        ), "Crop size must be the same for all transforms."
+        image_size = self.data_config.train_transform1.random_crop_size
+        self.byol_model = cast(
+            BYOL,
+            self.context.wrap_model(BYOL(net, image_size)),
+        )
+        self.byol_opt = self.context.wrap_optimizer(
+            build_byol_optimizer(self.hparams, self.byol_model)
+        )
+
+    def _init_classifiers(self) -> None:
+        """
+        Create classifier networks and optimizers, one for each candidate LR.
+        """
+        backbone_metadata = BACKBONE_METADATA_BY_NAME[self.hparams.backbone_name]
+        dataset_metadata = DATASET_METADATA_BY_NAME[self.data_config.dataset_name]
+        # Ensure learning rates are unique.
+        assert len(self.hparams.classifier.learning_rates) == len(
+            set(self.hparams.classifier.learning_rates)
+        )
+        self.cls_models = [
```

Comment:
later on, we loop through the train classifier dataloader and then loop through these models for each batch.  i wonder if it's better to loop through the lrs and then the dataloader to train one model at a time in which case we can just create a single classifier and model here and reset between lrs.

Reply:
Unfortunately, I can't see how to make this work.  All five trained classifiers are currently used in `evaluate_batch` after being trained in `on_validation_epoch_start`.  To keep only one model in memory, we'd have to be able to do the `on_validation_epoch_start` loop and the main `evaluate_batch` loop in between each reset.  Open to recommendations.

Reply:
I see, yeah what I suggested wouldn't work since we want to evaluate all 5 classifiers on the test set in the evaluate_batch loop.